internal/invocation/invocation_service.go

```diff
@@ -17,68 +17,147 @@
 
 import (
+	"context"
 	"fmt"
+	"sync/atomic"
+	"time"
 
+	"github.com/hazelcast/hazelcast-go-client/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/internal/cb"
 	ilogger "github.com/hazelcast/hazelcast-go-client/internal/logger"
 	"github.com/hazelcast/hazelcast-go-client/internal/proto"
 )
 
+const (
+	ready   = 0
+	stopped = 1
+)
+
 type Handler interface {
 	Invoke(invocation Invocation) (groupID int64, err error)
 }
 
 type Service struct {
-	requestCh       <-chan Invocation
-	urgentRequestCh <-chan Invocation
-	responseCh      <-chan *proto.ClientMessage
+	requestCh       chan Invocation
+	urgentRequestCh chan Invocation
+	responseCh      chan *proto.ClientMessage
 	// removeCh carries correlationIDs to be removed
-	removeCh    <-chan int64
+	removeCh    chan int64
 	doneCh      chan struct{}
 	invocations map[int64]Invocation
 	handler     Handler
 	logger      ilogger.Logger
+	state       int32
 }
 
 func NewService(
-	requestCh <-chan Invocation,
-	urgentRequestCh <-chan Invocation,
-	responseCh <-chan *proto.ClientMessage,
-	removeCh <-chan int64,
 	handler Handler,
 	logger ilogger.Logger) *Service {
 	service := &Service{
-		requestCh:       requestCh,
-		urgentRequestCh: urgentRequestCh,
-		responseCh:      responseCh,
-		removeCh:        removeCh,
+		requestCh:       make(chan Invocation),
+		urgentRequestCh: make(chan Invocation),
+		responseCh:      make(chan *proto.ClientMessage),
+		removeCh:        make(chan int64),
 		doneCh:          make(chan struct{}),
 		invocations:     map[int64]Invocation{},
 		handler:         handler,
 		logger:          logger,
+		state:           ready,
 	}
 	go service.processIncoming()
 	return service
 }
 
 func (s *Service) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.state, ready, stopped) {
+		return
+	}
 	close(s.doneCh)
+	close(s.requestCh)
+	close(s.urgentRequestCh)
+	close(s.responseCh)
+	close(s.removeCh)
 }
 
 func (s *Service) SetHandler(handler Handler) {
 	s.handler = handler
 }
 
+func (s *Service) SendRequest(ctx context.Context, inv Invocation) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("sending invocation: %w", ctx.Err())
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("sending urgent invocation: %w", hzerrors.ErrClientNotActive))
+	case s.requestCh <- inv:
+		return nil
+	}
+}
+
+func (s *Service) SendUrgentRequest(ctx context.Context, inv Invocation) error {
+	if atomic.LoadInt32(&s.state) != ready {
+		return cb.WrapNonRetryableError(hzerrors.ErrClientNotActive)
+	}
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("sending urgent invocation: %w", ctx.Err())
+	case <-s.doneCh:
+		return cb.WrapNonRetryableError(fmt.Errorf("sending urgent invocation: %w", hzerrors.ErrClientNotActive))
+	case s.urgentRequestCh <- inv:
+		return nil
+	}
+}
+
+func (s *Service) SendResponse(msg *proto.ClientMessage) error {
```

Comment:
This method name is confusing to me. 
It sounds like we are the server and sending the response to the client. 
Can we make it something like `ReceiveResponse` /`SetResponse`

Reply:
This method writes to the response channel. How about `WriteResponse`?

Reply:
Changed it to `WriteResponse` at: https://github.com/hazelcast/hazelcast-go-client/pull/655/commits/1ce567c695eccf2c344dcb5a41a51dd2a0de1218